---
title: "Data Preprocessing"
author: "Crystal Zang"
date: '2022-09-26'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Public on Feb 27, 2013
Title: Effect of sleep restriction on the human transcriptome during extended wakefulness	
	
Summary: In a cross balanced design human subjects were given one week of sufficient sleep and insufficient sleep (6 h per day). Following each of these conditions they were then kept awake for a day, a night and the subsequent day. During each of these periods of extended wakefulness 10 blood samples were taken, RNA was extracted from leukocytes, labelled and hybridised to human whole-genome microarrays

Design: A total of 438 samples comprising 26 human subjects, for which 20 samples across multiple time-points/sleep condition were collected.


```{r message=FALSE, warning=FALSE}
library(GEOquery)
library(dplyr)
library(tidyr)
library(Biobase)
library(expss)
library(matrixStats)

moller <- getGEO("gse39445") 
moller <- moller[["GSE39445_series_matrix.txt.gz"]]

archer <- getGEO("gse48113")
archer <- archer[["GSE48113_series_matrix.txt.gz"]]



process <- function(gse, sd.cutoff=0.5){
  if (sum(is.na(exprs(gse))) > 0){
    print("Missing value imputation using minimum gene expression.")
    rmin <- matrixStats::rowMins(exprs(gse), na.rm = TRUE)
    ij <- which(is.na(exprs(gse)), arr.ind = T)
    exprs(gse)[ij] <- rmin
  }else {
    print("No missing value.")
  }

  #gene filtering 
  sd <- matrixStats::rowSds(exprs(gse), na.rm=T)
  sd[is.na(sd)] <- 0.01
  gse.new <- gse[sd > sd.cutoff,]
  print(paste(dim(gse.new)[1], "genes left after filtering"))
  
  return(gse.new)
}


```


```{r}
#expression matrix. 
#col: sample
#row: gene
MAobject = as.matrix(trte.1)
dim(MAobject)

meta <- trte@featureData@data

### all variabls in trte data
varLabels(archer)

#circadian phase
table(trte$'circadianphase:ch1')

#treatment types
table(trte$'sleepprotocol:ch1')

#subject id
table(trte$'subject:ch1')

subjects_sub <- subjects%>%
  select(geo_accession, characteristics_ch1, 'sleepprotocol:ch1')


colnames(subjects_sub) <- c("sample_id", "subject_id", "tmt")
table(subjects_sub$tmt)
```

```{r}
# the data is already in log scale
# plot(density(exprs(trte.1),na.rm=T),main="density plot") # a density plots pooling all observations
library(limma)
# check expression distributions across samples 
plotDensities(exprs(trte), main = "Density plots for individuals", legend = F) 
```


### Heatmap by subjects

```{r}
# make a simple filtering: select genes with sd larger than 1
moller_c <- process(moller, 1.5)
#moller_c<-moller_c[,order(moller_c$'sleepprotocol:ch1')] 
#moller_c$'sleepprotocol:ch1'

dim(exprs(moller_c))

table(moller_c$'circadianphase:ch1') #44

moller_c.t_12 <-  moller_c[,moller_c$'circadianphase:ch1'== "-12"]
moller_c.t_9 <-  moller_c[,moller_c$'circadianphase:ch1'== "-9"]
moller_c.t_6 <-  moller_c[,moller_c$'circadianphase:ch1'== "-6"]
moller_c.t_3 <-  moller_c[,moller_c$'circadianphase:ch1'== "-3"]
moller_c.t0 <-  moller_c[,moller_c$'circadianphase:ch1'== "0"]
moller_c.t3 <-  moller_c[,moller_c$'circadianphase:ch1'== "3"]
moller_c.t6 <-  moller_c[,moller_c$'circadianphase:ch1'== "6"]
moller_c.t9 <-  moller_c[,moller_c$'circadianphase:ch1'== "9"]
moller_c.t12 <-  moller_c[,moller_c$'circadianphase:ch1'== "12"]

library("pheatmap")

#my_sample_col <- data.frame(moller_c.t_12$`sleepprotocol:ch1`)
#rownames(my_sample_col) <- colnames(exprs(moller_c.t_12))

#test
#png(file="fig/moller/check.png",width=1000, height=1500)
#pheatmap(exprs(moller_c.t_12), annotation_col = my_sample_col , cutree_cols = 2)
#dev.off()



#test 2 gene correlation
# png(file="/Users/czang/Documents/2069Omics/BIOST-2069-Project/fig/moller/genes_heatmap_t0.png",width=1500, height=1500)
# pheatmap(cor(t(exprs(moller_c.t0))))
# dev.off()
# 
# png(file="/Users/czang/Documents/2069Omics/BIOST-2069-Project/fig/moller/sample_heatmap_t0.png",width=1000, height=1000)
# pheatmap(cor(exprs(moller_c.t0)))
# dev.off()
```


# de
```{r}
library(samr)

table(moller_c.t0$'sleepprotocol:ch1')

GROUP = as.numeric(factor(moller_c.t0$`sleepprotocol:ch1`))


# Bioconductor requires data to stored in the S4 object
#create data object for sam
sam.data<-list(x=exprs(moller_c.t0),
               y=GROUP ,
               #geneid= 1:nrow(exprs(moller_c.t0)), 
              # genenames= row.names(exprs(moller_c.t0)),
              logged2=T
               )


#Perform SAM
samr.obj <- samr(sam.data, resp.type="Two class unpaired", nperms=100, random.seed=1)

delta.table <- samr.compute.delta.table(samr.obj)

siggenes.table<-samr.compute.siggenes.table(samr.obj, del=0, sam.data, delta.table)
str(siggenes.table)

a<-siggenes.table$genes.up
up <- a[as.numeric(a[,8])<1, ] #column qvalue: FDR, We want to control FDR at 1%
nrow(up) #Total number of up-regulated genes: 456; FDR=1%
up[1:10,]# top 10 eup-regulated genes

a<-siggenes.table$genes.lo
lo<-a[as.numeric(a[,8])<1, ]
nrow(lo) #Total number of down-regulated genes: 476; FDR=1%
lo[1:10,]# top 10 up-regulated genes

#Now let us draw a heatmap of the upregulated and down regulated genes. First, generate the matrix.
heatmap.data<-sam.data$x[as.numeric(c(up[,3], lo[,3])),]
heatmap(heatmap.data, Rowv=NA, Colv=NA)

```

```{r}
png(file="fig/moller/gene_level/heatmap1.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_12))), 
         cutree_rows = 2,  cutree_cols = 2, show_colnames = F, show_rownames =F, main = "Study: Moller, Time: -12", fontsize= 30)
dev.off() 

png(file="fig/moller/gene_level/heatmap2.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_9))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -9", fontsize= 30)
dev.off()

png(file="fig/moller/gene_level/heatmap3.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_6))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -6", fontsize= 30)
dev.off()

png(file="fig/moller/gene_level/heatmap4.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_3))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -3", fontsize= 30)
dev.off()

png(file="fig/moller/gene_level/heatmap5.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t0))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 0", fontsize= 30)
dev.off() 

png(file="fig/moller/gene_level/heatmap6.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t3))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 3", fontsize= 30)
dev.off() 

png(file="fig/moller/gene_level/heatmap7.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t6))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 6", fontsize= 30)
dev.off()

png(file="fig/moller/gene_level/heatmap8.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t9))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 9", fontsize= 30)
dev.off()

png(file="fig/moller/gene_level/heatmap9.png",
width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t12))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 12", fontsize= 30)
dev.off()
```

```{r}
#quantile normalization
#BiocManager::install("preprocessCore")
library(preprocessCore)
exprs(trte.1) <- normalize.quantiles(exprs(trte.1),copy=TRUE)

# check expression distributions acrosss samples 
limma::plotDensities(exprs(trte.1), main = "Density plots after Quantile normalization (TrTe)", legend = F) #the distribution of gene expression for all samples
boxplot(exprs(trte.1))
```


```{r message=FALSE, warning=FALSE}

archer_c = process(archer, 3)

sum(is.na(exprs(archer_c)))

dim(exprs(archer_c))
View(exprs(archer_c))
### all variabls in trte data
varLabels(archer_c)

#circadian phase
table(archer_c$'time point:ch1')


table(archer_c$'time point:ch1') #44

archer_c.t1 <-  archer_c[,archer_c$'time point:ch1'== "1"]
archer_c.t2<-  archer_c[,archer_c$'time point:ch1'== "2"]
archer_c.t3 <-  archer_c[,archer_c$'time point:ch1'== "3"]
archer_c.t4 <-  archer_c[,archer_c$'time point:ch1'== "4"]
archer_c.t5 <-  archer_c[,archer_c$'time point:ch1'== "5"]
archer_c.t6 <-  archer_c[,archer_c$'time point:ch1'== "6"]
archer_c.t7 <-  archer_c[,archer_c$'time point:ch1'== "7"]

sum(is.na(exprs(archer_c.t1)))
library("pheatmap")

png(file="fig/archer/gene_level/heatmap1.png",width=1000, height=1000)
pheatmap(cor(t(exprs(archer_c.t1))), 
         cutree_rows = 2,  cutree_cols = 2, show_colnames = F, show_rownames =F, main = "Study: Archer, Time: 1", fontsize= 30)
dev.off() 

png(file="fig/archer/gene_level/heatmap2.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_9))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -9", fontsize= 30)
dev.off()

png(file="fig/archer/gene_level/heatmap3.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_6))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -6", fontsize= 30)
dev.off()

png(file="fig/archer/gene_level/heatmap4.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t_3))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: -3", fontsize= 30)
dev.off()

png(file="fig/archer/gene_level/heatmap5.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t0))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 0", fontsize= 30)
dev.off() 

png(file="fig/archer/gene_level/heatmap6.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t3))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 3", fontsize= 30)
dev.off() 

png(file="fig/archer/gene_level/heatmap7.png",width=1000, height=1000)
pheatmap(cor(t(exprs(moller_c.t6))), cutree_rows = 2,  cutree_cols = 2, show_rownames = F, show_colnames = F, main = "Study: Moller, Time: 6", fontsize= 30)
dev.off()
```

